src/main/scala/tech/cryptonomic/conseil/tezos/TezosNodeOperator.scala

```diff
@@ -149,96 +150,66 @@ class TezosNodeOperator(val node: TezosRPCInterface)(implicit executionContext:
       headHash = blockHead.metadata.hash
       blocks <-
         if (headLevel <= maxLevel) Future.successful(List.empty)
-        else fetchBoundedBlockChain(network, blockHead, maxLevel + 1, headLevel, followFork)
+        else getBlocks(network, maxLevel + 1, headLevel, headHash, followFork)
     } yield blocks
 
   /**
-    * Fetches metadata for a block only, without waiting for the answer
-    * @param network   Which Tezos network to go against
-    * @param hash      Hash of the block
-    * @return          the metadata wrapped in a [[Future]]
+    * Get the blocks in a specified range.
+    * @param network Which tezos network to go against
+    * @param minLevel Minimum block level
+    * @param maxLevel Maximum block level
+    * @param startBlockHash If specified, start from the supplied block hash, otherwise, head of the chain.
+    * @param followFork If predecessor of the minLevel block appears to be on a fork, also capture the blocks on the fork.
+    * @return Blocks
     */
-  private def getMetadata(network: String, hash: String): Future[BlockMetadata] =
-    node.runAsyncGetQuery(network, s"blocks/$hash") map fromJson[BlockMetadata]
-
-  /**
-    * Asynchronously fetches a whole chain of metadata for blocks, using the parameters to
-    * decide when to stop the recursive descent on predecessors
-    *
-    * @param network    which Tezos network to go against
-    * @param current    The current metadata object to start descending
-    * @param minLevel   The lowest level for which to include results, unless a fork should be followed
-    * @param maxLevel   The highest level for which to include results
-    * @param followFork If the predecessor of the minLevel block appears to be on a fork, also capture the blocks on the fork.
-    * @param collected  The currently accumulated metadata so far
-    * @return           The metadata list, wrapped in a [[Future]]
-    */
-  def fetchChainedMetadata(
-    network: String,
-    current: BlockMetadata,
-    minLevel: Int,
-    maxLevel: Int,
-    followFork: Boolean,
-    collected: List[BlockMetadata] = List.empty
-  ) : Future[List[BlockMetadata]] = {
-    val currentLevel = current.header.level
```

Comment:
how comes all the following logic disappeared completely?

Reply:
That would require a lot of explanation, but the gist is that I spent a lot of time simplifying this algorithm by separating the part that deals with forks and the part that doesn't deal with forks. Once my forking logic is done, the result will be equivalent, and hopefully a lot more obvious as to why the algorithm is correct.

Reply:
Plus, this part is unaware of the recent fact we have learned that we can use offsets from the head using rpc calls.